core/trino-main/src/main/java/io/trino/execution/scheduler/HashDistributionSplitAssigner.java

```diff
@@ -183,20 +221,59 @@ private static Map<Integer, TaskPartition> createOutputPartitionToTaskPartition(
         OutputDataSizeEstimate mergedEstimate = OutputDataSizeEstimate.merge(partitionedSourcesEstimates);
         ImmutableMap.Builder<Integer, TaskPartition> result = ImmutableMap.builder();
         PriorityQueue<PartitionAssignment> assignments = new PriorityQueue<>();
-        assignments.add(new PartitionAssignment(new TaskPartition(), 0));
-        for (int outputPartitionId = 0; outputPartitionId < partitionCount; outputPartitionId++) {
-            long outputPartitionSize = mergedEstimate.getPartitionSizeInBytes(outputPartitionId);
-            if (assignments.peek().assignedDataSizeInBytes() + outputPartitionSize > targetPartitionSizeInBytes
-                    && assignments.size() < partitionCount) {
-                assignments.add(new PartitionAssignment(new TaskPartition(), 0));
+        for (int partitionId = 0; partitionId < partitionCount; partitionId++) {
+            long partitionSizeInBytes = mergedEstimate.getPartitionSizeInBytes(partitionId);
+            if (assignments.isEmpty() || assignments.peek().assignedDataSizeInBytes() + partitionSizeInBytes > targetPartitionSizeInBytes || !canMerge) {
+                TaskPartition taskPartition = createTaskPartition(
+                        partitionSizeInBytes,
+                        targetPartitionSizeInBytes,
+                        partitionedSources,
+                        outputDataSizeEstimates,
+                        partitionId,
+                        canSplit);
+                result.put(partitionId, taskPartition);
+                assignments.add(new PartitionAssignment(taskPartition, partitionSizeInBytes));
+            }
+            else {
+                PartitionAssignment assignment = assignments.poll();
+                result.put(partitionId, assignment.taskPartition());
+                assignments.add(new PartitionAssignment(assignment.taskPartition(), assignment.assignedDataSizeInBytes() + partitionSizeInBytes));
             }
-            PartitionAssignment assignment = assignments.poll();
-            result.put(outputPartitionId, assignment.taskPartition());
-            assignments.add(new PartitionAssignment(assignment.taskPartition(), assignment.assignedDataSizeInBytes() + outputPartitionSize));
         }
         return result.buildOrThrow();
     }
 
+    private static TaskPartition createTaskPartition(
+            long partitionSizeInBytes,
+            long targetPartitionSizeInBytes,
+            Set<PlanNodeId> partitionedSources,
+            Map<PlanNodeId, OutputDataSizeEstimate> outputDataSizeEstimates,
```

Comment:
Currently we always have estimates. Will that break apart when we start doing speculative execution when estimates are not always there. Or will they still be available even then?

Reply:
That's a good question. 

Basically there are two possible scenarios:

1. Some upstream tasks have finished. In this scenario we can project estimates by adjusting for the tasks that are still in process. This will usually be the case when starting a hash distributed stage that reads data from a source distributed stage (e.g.: TableScan)
2. Non upstream tasks have finished. This scenario is more tricky. I guess it would generally indicate that the number of input tasks is small (e.g.: less than available slots in the cluster). In such scenario i guess we would have to resort to simply telling the HashDistributionSplitAssigner to create a fixed number of tasks with uniform partition assignment. It should in theory be good enough, as generally it would indicate that the input data set is not "large". However if it's endup being problematic we would probably need to improve the estimate reporting mechanism to allow incomplete tasks to report intermediate "estimates".